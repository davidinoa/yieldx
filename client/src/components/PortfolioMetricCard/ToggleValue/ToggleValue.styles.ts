import styled from "styled-components";
import { mediaQuery } from "../../../theme/breakpoints.style";

export const ToggleLink = styled.button`
  margin-left: auto;
  color: var(--lightest);
  font-size: 0.8rem;
  text-decoration: underline;
  padding: 0;
  background: none;
  border: none;
`;

const toggleButtonSize = 24;
const toggleWrapperPadding = 4;
const factor = 0.75;

export const ToggleSwitch = styled.div`
  position: relative;
  background: var(--page);
  border-radius: ${toggleButtonSize * factor}px;
  height: ${(toggleButtonSize + toggleWrapperPadding * 2) * factor}px;

  ${mediaQuery("xxl")} {
    border-radius: ${toggleButtonSize}px;
    height: ${toggleButtonSize + toggleWrapperPadding * 2}px;
  }

  .toggle-button {
    background: none;
    z-index: 1;
    border: none;
    padding: 7px;
    height: ${toggleButtonSize * factor}px;
    ${mediaQuery("xxl")} {
      height: ${toggleButtonSize}px;
    }
    font-weight: bold;
    font-size: 0.9rem;
    line-height: 12px;
    letter-spacing: 0.1em;
    text-transform: uppercase;
  }
  &::before {
    content: "";
    background: var(--light);
    width: 50%;
    position: absolute;
    border-radius: 12px;
    transition: all 200ms ease-out;
    top: ${toggleWrapperPadding * factor}px;
    height: ${toggleButtonSize * factor}px;
    left: 2px;
    ${mediaQuery("xxl")} {
      top: ${toggleWrapperPadding}px;
      height: ${toggleButtonSize}px;
    }
  }

  &.toggled {
    &::before {
      transition: all 200ms ease-out;
      left: 47%;
    }
  }
`;
