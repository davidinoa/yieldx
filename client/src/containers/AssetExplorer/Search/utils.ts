interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
