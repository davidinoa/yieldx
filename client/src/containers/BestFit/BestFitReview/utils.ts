import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
