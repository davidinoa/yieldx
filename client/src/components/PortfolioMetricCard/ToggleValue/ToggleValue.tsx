import React from "react";
import _ from "lodash";
import clsx from "clsx";
import { ToggleLink, ToggleSwitch } from "./ToggleValue.styles";

interface Props {
  switchToggleLabels?: [string, string];
  onToggle: () => void;
  current: string;
  className?: string;
}

export default function ToggleValue({
  onToggle,
  switchToggleLabels,
  current,
  className
}: Props) {
  if (switchToggleLabels) {
    const [firstLabel, secondLabel] = switchToggleLabels;
    return (
      <ToggleSwitch
        className={clsx(
          `d-flex align-items-center ${
            className !== undefined ? className : ""
          }`,
          {
            toggled: current === secondLabel
          }
        )}
      >
        <button
          type="button"
          data-testid="ToggleValue-Percent"
          className="d-flex justify-content-center align-items-center toggle-button"
          onClick={onToggle}
        >
          {firstLabel}
        </button>
        <button
          type="button"
          data-testid="ToggleValue-Dollar"
          className="d-flex justify-content-center align-items-center toggle-button"
          onClick={onToggle}
        >
          {secondLabel}
        </button>
      </ToggleSwitch>
    );
  }

  return (
    <ToggleLink type="button" onClick={onToggle}>
      {_.startCase(current)}
    </ToggleLink>
  );
}
