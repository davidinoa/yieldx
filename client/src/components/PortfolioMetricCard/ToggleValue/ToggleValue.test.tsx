import React from "react";
import userEvent from "@testing-library/user-event";
import { render, screen } from "../../../test/utils";
import ToggleValue from "./ToggleValue";

describe("ToggleValue", () => {
  test("Should display text button with toggleLabel prop", () => {
    const mockToggle = jest.fn();
    const { asFragment } = render(
      <ToggleValue onToggle={mockToggle} current="Annual" />
    );
    const toggleBtn = screen.getByText(/Annual/i);
    expect(toggleBtn).toBeInTheDocument();
    userEvent.click(toggleBtn);
    expect(mockToggle).toHaveBeenCalled();
    expect(asFragment()).toMatchSnapshot();
  });
  test("Should render switch if label is $ or %", () => {
    const mockToggle = jest.fn();
    const { asFragment } = render(
      <ToggleValue
        onToggle={mockToggle}
        current="%"
        switchToggleLabels={["%", "$"]}
      />
    );
    const toggleBtn = screen.getByTestId("ToggleValue-Percent");
    expect(toggleBtn).toBeInTheDocument();
    userEvent.click(toggleBtn);
    expect(mockToggle).toHaveBeenCalled();
    expect(asFragment()).toMatchSnapshot();
  });
});
