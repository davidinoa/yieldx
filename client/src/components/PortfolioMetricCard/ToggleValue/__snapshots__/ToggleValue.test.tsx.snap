// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleValue Should display text button with toggleLabel prop 1`] = `
<DocumentFragment>
  .c0 {
  margin-left: auto;
  color: var(--lightest);
  font-size: 0.8rem;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  padding: 0;
  background: none;
  border: none;
}

<button
    class="c0"
    type="button"
  >
    Annual
  </button>
</DocumentFragment>
`;

exports[`ToggleValue Should render switch if label is $ or % 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  background: var(--page);
  border-radius: 18px;
  height: 24px;
}

.c0 .toggle-button {
  background: none;
  z-index: 1;
  border: none;
  padding: 7px;
  height: 18px;
  font-weight: bold;
  font-size: 0.9rem;
  line-height: 12px;
  -webkit-letter-spacing: 0.1em;
  -moz-letter-spacing: 0.1em;
  -ms-letter-spacing: 0.1em;
  letter-spacing: 0.1em;
  text-transform: uppercase;
}

.c0::before {
  content: "";
  background: var(--light);
  width: 50%;
  position: absolute;
  border-radius: 12px;
  -webkit-transition: all 200ms ease-out;
  transition: all 200ms ease-out;
  top: 3px;
  height: 18px;
  left: 2px;
}

.c0.toggled::before {
  -webkit-transition: all 200ms ease-out;
  transition: all 200ms ease-out;
  left: 47%;
}

@media (min-width:1600px) {
  .c0 {
    border-radius: 24px;
    height: 32px;
  }
}

@media (min-width:1600px) {
  .c0 .toggle-button {
    height: 24px;
  }
}

@media (min-width:1600px) {
  .c0::before {
    top: 4px;
    height: 24px;
  }
}

<div
    class="c0 d-flex align-items-center "
  >
    <button
      class="d-flex justify-content-center align-items-center toggle-button"
      data-testid="ToggleValue-Percent"
      type="button"
    >
      %
    </button>
    <button
      class="d-flex justify-content-center align-items-center toggle-button"
      data-testid="ToggleValue-Dollar"
      type="button"
    >
      $
    </button>
  </div>
</DocumentFragment>
`;
